bpo-31985: Deprecate aifc.openfp

briancurtin opened the pull request:
aifc.openfp had pointed to aifc.open since 1993* and it was both undocumented and untested since being assigned as a matter of backwards compatibility. This change begins its formal deprecation.

\* https://github.com/python/cpython/commit/7bc817d5ba917528e8bd07ec461c635291e7b06a

<!-- issue-number: bpo-31985 -->
https://bugs.python.org/issue31985
<!-- /issue-number -->

briancurtin committed:
Deprecate aifc.openfp

aifc.openfp had pointed to aifc.open since 1993 and it was both
undocumented and untested since being assigned as a matter of backwards
compatibility. This change begins its formal deprecation.

berkerpeksag commented on Lib/test/test_aifc.py:
```diff
@@ -145,6 +146,20 @@ class AifcALAWTest(AifcTest, unittest.TestCase):
+            aifc.openfp(arg, mode=mode)
+
+            self.assertTrue(len(w) == 1)
+            self.assertTrue(issubclass(w[0].category, DeprecationWarning))
```
Style nit: `assertIsInstance` might be better option here.

briancurtin replied:
I initially thought that as well, but `category` isn't an instance there.

The other comments look good and I'll push a fix for them.

briancurtin replied:
By using `assertWarns` we get the behavior we needed from those two `assertTrue` lines, so I just removed them in ebe685d.

berkerpeksag commented on Lib/test/test_aifc.py:
```diff
@@ -145,6 +146,20 @@ class AifcALAWTest(AifcTest, unittest.TestCase):
+            warnings.simplefilter("always")
+            aifc.openfp(arg, mode=mode)
+
+            self.assertTrue(len(w) == 1)
```
Style nit: I'd use `assertEqual` instead.

berkerpeksag commented on Lib/test/test_aifc.py:
```diff
@@ -145,6 +146,20 @@ class AifcALAWTest(AifcTest, unittest.TestCase):
+        arg = "arg"
+        mode = "mode"
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
```
We could replace those two lines with `assertWarns`.

berkerpeksag reviewed (commented):
LGTM, I just left some minor comments.

briancurtin committed:
Switch to using assertWarns

By using assertWarns we no longer need to check the number of warnings or
the category of the warning, as assertWarns already handles that for us.

briancurtin committed:
Add sunau and wave deprecations for openfp

This additionally moves the test to Lib/audiotests.py and has MiscTests
in each module's tests inherit from the new AudioMiscTests which does
the deprecation test.

serhiy-storchaka commented on Lib/aifc.py:
```diff
@@ -915,7 +915,10 @@ def open(f, mode=None):
         raise Error("mode must be 'r', 'rb', 'w', or 'wb'")
 
-openfp = open # B/W compatibility
+def openfp(f, mode=None):
```
Please add a comment that `test_pyclbr.py` should be updated after removing `openfp`.

briancurtin replied:
Comment added. After this merges I'm going to create an issue in Roundup and assign it to myself so it gets formally tracked.

serhiy-storchaka reviewed (commented):
Thank you Brian!

Please replace `wave.openfp` with `wave.open` in `sndhdr.py`. Add `versiondeprecated` in the documentation for `wave.openfp` and `sunau.openfp`.

briancurtin committed:
Added versiondeprecated in sunau and wave docs

This additionally adds a TODO in test_pyclbr around using openfp, though
it shouldn't be changed until after 3.7.

serhiy-storchaka commented on Doc/library/sunau.rst:
```diff
@@ -63,6 +63,8 @@ The :mod:`sunau` module defines the following functions:
 
    A synonym for :func:`.open`, maintained for backwards compatibility.
 
+   .. versiondeprecated:: 3.7
```
Oh, sorry, the correct name of this directive is `deprecated`! Or `deprecated-removed` if you have assigned the date of removing.

briancurtin committed:
Update deprecation directive to deprecated-removed

These are being deprecated now for 3.7 and will be removed in 3.9.
